[Run Plugin Manager] Update additional options

davidegiacometti opened the pull request:
## Summary of the Pull Request

**What is this about:**
Add/Remove additional options for PT Run plugins on startup

**What is include in the PR:** 

**How does someone test / validate:** 
Need to be tested in debug.
- Add these additional options to Indexer Plugin

```
new PluginAdditionalOption()
{
    Key = "Option0",
    DisplayLabel = "Option 0",
    Value = false,
},
new PluginAdditionalOption()
{
    Key = "Option1",
    DisplayLabel = "Option 1",
    Value = false,
},
```

- Build and run PT
- Option 0 and Option 1 should be displayed in Plugin Manager
- PT Run should work
- Close PT
- Remove Option 0 and Option 1 from the additional options
- Build and run PT
- Option 0 and Option 1 should not be displayed in Plugin Manager

## Quality Checklist

- [x] **Linked issue:** #12483 #9929
- [ ] **Communication:** I've discussed this with core contributors in the issue. 
- [ ] **Tests:** Added/updated and all pass
- [ ] **Installer:** Added/updated and all pass
- [ ] **Localization:** All end user facing strings can be localized
- [ ] **Docs:** Added/ updated
- [x] **Binaries:** Any new files are added to WXS / YML
   - [x] No new binaries
   - [ ] [YML for signing](https://github.com/microsoft/PowerToys/blob/master/.pipelines/pipeline.user.windows.yml#L68) for new binaries
   - [ ] [WXS for installer](https://github.com/microsoft/PowerToys/blob/master/installer/PowerToysSetup/Product.wxs) for new binaries

## Contributor License Agreement (CLA)
A CLA must be signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/PowerToys) and sign the CLA.

mykhailopylyp commented on src/modules/launcher/PowerLauncher/SettingsReader.cs:
```diff
@@ -215,5 +216,37 @@ private static IEnumerable<PowerLauncherPluginSettings> GetDefaultPluginsSetting
+
+                if (plugin is ISettingProvider && (plugin as ISettingProvider).AdditionalOptions != null)
+                {
+                    var pluginSettings = settings.Plugins.Single(p => p.Id == pluginPair.Metadata.ID);
```
Let's change it to SingleOrDefault. It is possible that newly added plugin is not in user settings yet.

mykhailopylyp commented:
The main concern is that we don't keep update user settings with default options centralized. Right now we have two methods that are responsible for it: `CombineWithDefaultSettings` and `UpdateAdditionalOptionsSettings`. So
maybe we should move `UpdateAdditionalOptionsSettings` into `CombineWithDefaultSettings`?

davidegiacometti commented:
@mykhailopylyp I understand.
I have trying to combine the methods and I figured out that this logic is causing this issue: https://github.com/microsoft/PowerToys/issues/12921
The point is that additional settings should be updated once on PT Run startup and not in the `ReadSettings()` that is used by the file system watcher. Does it make sense to move `UpdateAdditionalOptionsSettings` and `CombineWithDefaultSettings` outside `ReadSettings()`?

mykhailopylyp commented:
>I have trying to combine the methods and I figured out that this logic is causing this issue: #12921
Can you elaborate here?

>The point is that additional settings should be updated once on PT Run startup and not in the ReadSettings() that is used by the file system watcher.

Actually, `CombineWithDefaultSettings` should be called only on PT Run startup as well.

>Does it make sense to move UpdateAdditionalOptionsSettings and CombineWithDefaultSettings outside ReadSettings()?

Don't see any reasons why should we move it out.

davidegiacometti commented:
Add new plugins, add missing additional options and update additional options labels should be done only once at startup.
Doing this inside the file system watcher is causing this issue: https://github.com/microsoft/PowerToys/issues/12921
Since every time the file is changed by the settings, the launcher is going to update the additional options and save the file again.

mykhailopylyp commented:
>Add new plugins, add missing additional options and update additional options labels should be done only once at startup.
Doing this inside the file system watcher is causing this issue: #12921

In this PR, `UpdateAdditionalOptionsSettings` is called from ReadSettings so it is called on file change as well.
Let's move `CombineWithDefaultSettings` options and and `UpdateAdditionalOptionsSettings` into separate method that is called only on startup.

davidegiacometti commented:
@mykhailopylyp I have updated the PR. Let me know what do you think about this solution.

htcfreek commented:
> @mykhailopylyp I have updated the PR. Let me know what do you think about this solution.
> Tested:
> - Add new plugins
> - Add new additional options
> - Update additional options labels
> - Remove additional options
> - Remove additional options if `ISettingProvider` is removed.
> 
> Further improvements could be:
> - Remove plugins (not sure if needed)
> - Sort additional options

Regarding the further improvements I am thinking on #9929 or at least on label changes (e.g. correcting misspelling) too.

mykhailopylyp commented on src/modules/launcher/PowerLauncher/SettingsReader.cs:
```diff
@@ -215,5 +195,52 @@ private static IEnumerable<PowerLauncherPluginSettings> GetDefaultPluginsSetting
+
+            foreach (var pluginPair in PluginManager.AllPlugins)
+            {
+                var plugin = pluginPair.Plugin;
+                var pluginSettings = settings.Plugins.SingleOrDefault(p => p.Id == pluginPair.Metadata.ID);
+
+                // Add plugin
+                if (pluginSettings == null)
+                {
+                    pluginSettings = plugins[pluginPair.Metadata.ID];
+                    settings.Plugins = settings.Plugins.Append(pluginSettings).OrderBy(p => p.Name);
+                    continue;
+                }
+
+                // Update additional options
+                if (plugin is ISettingProvider && (plugin as ISettingProvider).AdditionalOptions != null)
+                {
+                    foreach (var option in (plugin as ISettingProvider).AdditionalOptions)
+                    {
+                        var pluginSetting = pluginSettings.AdditionalOptions.SingleOrDefault(o => o.Key == option.Key);
+
+                        if (pluginSetting == null)
+                        {
+                            pluginSettings.AdditionalOptions = pluginSettings.AdditionalOptions.Append(option);
+                        }
+                        else
+                        {
+                            pluginSetting.DisplayLabel = option.DisplayLabel;
+                        }
+                    }
+
+                    var optionsToDelete = pluginSettings.AdditionalOptions.Except((plugin as ISettingProvider).AdditionalOptions);
+                    pluginSettings.AdditionalOptions = pluginSettings.AdditionalOptions.Except(optionsToDelete);
+                }
+                else if (!(plugin is ISettingProvider))
+                {
+                    pluginSettings.AdditionalOptions = Enumerable.Empty<PluginAdditionalOption>();
+                }
+            }
+        }
```
```suggestion
        private static void UpdateSettings(PowerLauncherSettings settings)
        {
            var defaultPlugins = GetDefaultPluginsSettings().ToDictionary(x => x.Id);
            foreach (PowerLauncherPluginSettings plugin in settings.Plugins)
            {
                if (defaultPlugins.ContainsKey(plugin.Id))
                {
                    var additionalOptions = CombineAdditionalOptions(defaultPlugins[plugin.Id].AdditionalOptions, plugin.AdditionalOptions);
                    defaultPlugins[plugin.Id] = plugin;
                    defaultPlugins[plugin.Id].AdditionalOptions = additionalOptions;
                }
            }

            settings.Plugins = defaultPlugins.Values.ToList();
        }

        private static IEnumerable<PluginAdditionalOption> CombineAdditionalOptions(IEnumerable<PluginAdditionalOption> defaultAdditionalOptions, IEnumerable<PluginAdditionalOption> additionalOptions)
        {
            var defaultOptions = defaultAdditionalOptions.ToDictionary(x => x.Key);
            foreach (var option in additionalOptions)
            {
                if (defaultOptions.ContainsKey(option.Key))
                {
                    defaultOptions[option.Key].DisplayLabel = option.DisplayLabel;
                    defaultOptions[option.Key].Value = option.Value;
                }
            }

            return defaultOptions.Values;
        }
```

mykhailopylyp replied:
Even though the previous implementation is correct I think that the proposed one is easier to understand

davidegiacometti commented:
@mykhailopylyp I have updated the PR. Now it looks cleaner!
Since this solution is not using `PluginAdditionalOption` comparison I have removed changes made to `PluginAdditionalOption.cs`.
To fully support https://github.com/microsoft/PowerToys/issues/9929 I have tweaked the code a bit to update plugin name, description and additional options label.

htcfreek commented:
> To fully support https://github.com/microsoft/PowerToys/issues/9929 I have tweaked the code a bit to update plugin name, description and additional options label. 

Sounds good. The only thing I don't know is if the values are updated in json if language is changed. 🤔 But this could be added by an other PR to PT Run itself.

davidegiacometti commented:
> > To fully support #9929 I have tweaked the code a bit to update plugin name, description and additional options label.
> 
> Sounds good. The only thing I don't know is if the values are updated in json if language is changed. 🤔 But this could be added by an other PR to PT Run itself.

Yes, plugin name, description and options are always updated on PT startup.

mykhailopylyp reviewed (approved):
LGTM! Thank you for your contribution

htcfreek commented:
@jaimecbernardo 
This issue needs to be merged together with #12939. (Otherwise the settings won't be updated in ux.)

jaimecbernardo commented:
@htcfreek ,
Done. Thank you for the contribution!

htcfreek commented:
> @davidegiacometti ,
> Done. Thank you for the contribution!
